### PR TITLE
PM-9682: Verify with PIN on add edit view

### PR DESCRIPTION
**Pull request by shannon-livefront**
## 🎟️ Tracking

[PM-9682](https://bitwarden.atlassian.net/browse/PM-9682)

## 📔 Objective

If the user's vault is unlocked, their device has no lock screen configured, and they have a Bitwarden PIN configured, allow them to authenticate with their PIN when saving a passkey on the vault add edit view.

## 📸 Screenshots

| Before | After |
| --- | --- |
| <video src="https://github.com/user-attachments/assets/6af9ead6-e206-4833-8fba-b68685a2e9ef" /> | <video src="https://github.com/user-attachments/assets/5df5e553-8873-466b-a6c5-c14783a0c997" /> |

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

[PM-9682]: https://bitwarden.atlassian.net/browse/PM-9682?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/addedit/VaultAddEditScreen.kt [hunk 1]
```diff
@@ -398,6 +416,23 @@ private fun VaultAddEditItemDialogs(
             )
         }
 
+        is VaultAddEditState.DialogState.Fido2PinPrompt -> {
+            BitwardenPinDialog(
+                onConfirmClick = { onSubmitPinFido2Verification(it) },
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/addedit/VaultAddEditScreen.kt:421]: Can we drop the outer lambda:

```kotlin
onConfirmClick = onSubmitPinFido2Verification
```

**Review by david-livefront — APPROVED** [review]

File: app/src/main/java/com/x8bit/bitwarden/ui/platform/components/dialog/BitwardenPinDialog.kt [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+ * Represents a Bitwarden-styled dialog for the user to enter their PIN.
+ *
+ * @param onConfirmClick called when the confirm button is clicked and emits the entered PIN.
+ * @param onDismissRequest called when the user attempts to dismiss the dialog (for example by
+ * tapping outside of it).
+ */
+@OptIn(ExperimentalComposeUiApi::class)
+@Composable
+fun BitwardenPinDialog(
+    onConfirmClick: (masterPassword: String) -> Unit,
```
Comment (dseverns-livefront) [inline app/src/main/java/com/x8bit/bitwarden/ui/platform/components/dialog/BitwardenPinDialog.kt:33]: ⛏️  maybe name this "pin" ?

File: app/src/main/java/com/x8bit/bitwarden/ui/platform/components/dialog/BitwardenPinDialog.kt [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    onDismissRequest: () -> Unit,
+) {
+    var pin by remember { mutableStateOf("") }
```
Comment (dseverns-livefront) [inline app/src/main/java/com/x8bit/bitwarden/ui/platform/components/dialog/BitwardenPinDialog.kt:36]: 💭 should this be saveable, the edge case of someone entering `**` -> `change to dark mode` and they would lose the values entered. Not sure if we are specifically not doing this as security thing.

File: app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/itemlisting/VaultItemListingViewModel.kt [hunk 4]
```diff
@@ -659,7 +700,7 @@ class VaultItemListingViewModel @Inject constructor(
     }
 
     private fun handleDismissDialogClick() {
-        mutableStateFlow.update { it.copy(dialogState = null) }
+        clearDialogState()
```
Comment (dseverns-livefront) [inline app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/itemlisting/VaultItemListingViewModel.kt:703]: 👍 👍

File: app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/addedit/VaultAddEditViewModel.kt [hunk 5]
```diff
@@ -1583,28 +1617,57 @@ class VaultAddEditViewModel @Inject constructor(
...
+                    )
                 }
+            }
+        }
+    }
+
+    private fun handleValidateFido2PinResultReceive(
+        action: VaultAddEditAction.Internal.ValidateFido2PinResultReceive,
+    ) {
+        mutableStateFlow.update { it.copy(dialog = null) }
```
Comment (SaintPatrck) [inline app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/addedit/VaultAddEditViewModel.kt:1634]:
```suggestion
        clearDialogState()
```

File: app/src/main/java/com/x8bit/bitwarden/ui/platform/components/dialog/BitwardenPinDialog.kt (hunk 3 above)
Comment (shannon-livefront) [reply app/src/main/java/com/x8bit/bitwarden/ui/platform/components/dialog/BitwardenPinDialog.kt:36]: Excellent question - I actually don't know so I'll defer to @david-livefront and @SaintPatrck for the security aspect of that question.

File: app/src/main/java/com/x8bit/bitwarden/ui/platform/components/dialog/BitwardenPinDialog.kt (hunk 2 above)
Comment (shannon-livefront) [reply app/src/main/java/com/x8bit/bitwarden/ui/platform/components/dialog/BitwardenPinDialog.kt:33]: Great catch, thank you!

File: app/src/main/java/com/x8bit/bitwarden/ui/platform/components/dialog/BitwardenPinDialog.kt (hunk 3 above)
Comment (SaintPatrck) [reply app/src/main/java/com/x8bit/bitwarden/ui/platform/components/dialog/BitwardenPinDialog.kt:36]: As a user, if my theme changes while this dialog is displayed I would not expect to lose my input. Beyond a user manually changing it, but still an edge case, the theme could change if it's based on time of day. 

@david-livefront this is reminiscent of the issues you worked on with Dillon regarding input being retained for too long. Any concerns about that here?

Comment (david-livefront) [reply app/src/main/java/com/x8bit/bitwarden/ui/platform/components/dialog/BitwardenPinDialog.kt:36]: I would expect us to not save this for security reasons, I believe we don't use `remeberSavable` in the `BitwardenMasterPasswordDialog` either.

It is annoying, the theme is an edge-case but rotation will also clear the state and we just have to live with that.

**Review by SaintPatrck — APPROVED** [review]

**Pull request merged**